src/chocolatey/infrastructure.app/services/ShimGenerationService.cs

```diff
@@ -87,34 +90,83 @@ public void install(ChocolateyConfiguration configuration, PackageResult package
+
+            if (noShims)
+            {
+                this.Log().Debug(() => "Automatic package shimming disabled for {0}".format_with(packageResult.Name));
+            }
+
+            if (existingShims.Count > 0)
+            {
+                this.Log().Debug(() => "Removing previous shims for {0}".format_with(packageResult.Name));
+                uninstall_shims(existingShims);
+            }
+
+            if (!noShims)
+            {
+                //gather all relevant .exes in the folder
+                var exeFiles = get_exe_files_to_shim(packageResult.InstallLocation);
+                install_shims(configuration, packageResult.Name, exeFiles);
+            }
+        }
+
+        public void uninstall(ChocolateyConfiguration configuration, PackageResult packageResult)
+        {
+            var existingShims = _registry.get_all(packageResult.Name);
+            uninstall_shims(existingShims);
+        }
+
+        public void take_snapshot()
+        {
+            _registry.create_snapshot();
+        }
+
+        /// <summary>
+        ///   Installs shimgens for the package
+        /// </summary>
+        /// <param name="configuration">The configuration.</param>
+        /// <param name="packageName">The package name.</param>
+        /// <param name="exeFiles">The exe files to shim</param>
+        private void install_shims(ChocolateyConfiguration configuration, string packageName, IList<string> exeFiles)
+        {
+            var args = ExternalCommandArgsBuilder.build_arguments(configuration, _shimGenArguments);
+
+            foreach (string file in exeFiles)
             {
-                if (_fileSystem.file_exists(file + ".ignore")) continue;
                 bool isGui = _fileSystem.file_exists(file + ".gui");
                 //todo: v2 be able to determine gui automatically
 
-                var args = ExternalCommandArgsBuilder.build_arguments(configuration, _shimGenArguments);
                 var shimLocation = _fileSystem.combine_paths(ApplicationParameters.ShimsLocation, _fileSystem.get_file_name(file));
                 var argsForPackage = args.Replace(PATH_TOKEN, file.Replace(ApplicationParameters.InstallLocation, "..\\")).Replace(OUTPUT_TOKEN, shimLocation).Replace(ICON_PATH_TOKEN, file);
                 if (isGui)
                 {
                     argsForPackage += " --gui";
                 }
 
+                // check for an existing shim
+                ShimRecord record = _registry.get(shimLocation);
+                if (record != null)
+                {
+                    this.Log().Error(() => " ShimGen cannot overwrite shim {0} in {1} package".format_with(_fileSystem.get_file_name(file), record.PackageName));
+                    this.Log().Debug(() => "  Shim: {0}{1}  Target: {2}{1}  New target: {3}{1}".format_with(shimLocation, Environment.NewLine, record.TargetFile, file));
+                    Environment.ExitCode = 1;

```

Comment:
I'm not sure shim generation would be one of those things I would count as causing an exit code of an error. I would see if enhanced exit codes are turned on and let's specify an exit code specific for this. As in a type of inconclusive installation.

Reply:
Okay. I'll use enhanced exit code 3 (since 2 has been used for _no results_)

Reply:
On second thoughts (or rather getting my head around it again):

> I'm not sure shim generation would be one of those things I would count as causing an exit code of an error.

I am actually copying existing behaviour, because an error exit code is specifically returned if shimgen.exe itself fails.

https://github.com/chocolatey/choco/blob/26ac2df953ea1c8af2c03c33525ac0600c8e46f6/src/chocolatey/infrastructure.app/services/ShimGenerationService.cs#L173-L176